### DISCUSSION TITLE
Canvas Renderer Support in v8

### DISCUSSION THREAD

**Discussion by AlshaerDEV** [pixijs/pixijs#10682] [v8 General] [answered]
Hello!
Canvas Renderer Support in v8
I'm currently using the pixi.js-legacy package, and I'm planning to upgrade to Pixi v8. Recently, I noticed in the `src/rendering/renderers/autoDetectRenderer.ts` file, there's no mention of a Canvas Renderer:

![autoDetectRenderer.ts screenshot](https://github.com/pixijs/pixijs/assets/52708327/33399297-52c4-44c7-9989-abb4ca64562e)

However, the README says this:

![README screenshot](https://github.com/pixijs/pixijs/assets/52708327/7d368bde-1fd9-4f17-b4b5-89212a3d7ded)

This has left me a bit confused. Could you clarify whether v8 includes a Canvas Renderer?

Thank you for your support!

**Comment by bigtimebuddy** [comment]
v8 does not have a CanvasRenderer yet but we are planning to re-add it. There was a lot to figure out with WebGPU and we didn't want to complicate it for ourselves. If you require canvas, continue to use v7 legacy bundle (pixi.js-legacy).

**Reply by AlshaerDEV** [reply]
Hi! Could you clarify whether v8 now includes a Canvas Renderer ?
Thank you for your support!

**Reply by bigtimebuddy** [reply]
Sorry, no updates yet on CanvasRenderer! We have budgeted funds to developing this feature, but have had a difficult time finding the right dev who can deliver it.

**Reply by Mistik** [reply]
> Sorry, no updates yet on CanvasRenderer! We have budgeted funds to developing this feature, but have had a difficult time finding the right dev who can deliver it.

Whats the budget? Maybe I can donate more towards it so you can find the right dev

**Comment by andrewschreiber** [comment]
Eager for v8 canvas renderer support!

Would it be possible to break down the goal into subtasks or outstanding architecture choices so the community could help?

**Comment by kirs26** [comment]
Eager for v8 canvas renderer support!

**Comment by CrushedPixel** [comment]
Also waiting for v8 CanvasRenderer support - I have built myself into a corner in my project where I am spawning too many WebGL contexts for browsers to handle (limit is 16), so being able to use CanvasRenderer for some of the less demanding rendering contexts would be big.

**Reply by joshstrike** [reply]
+1. This is a significant danger with something like Pixi, although it was also a danger back when Java embeds or Flash embeds were a thing on the web. I don't like having the limit, because sometimes you need a LOT of separate charts and things. But having 16 separate canvases animating will not do your users a lot more favors than having 16 WebGL contexts, either. Mixing black-box bullets with regular HTML has always been an architectural problem - and a very understandable one when you want to have lots of cool modules going in separate places, but don't want to make your whole site a black box. I'm not a typical engineer, and 99% would disagree with me, but I think if you need that many custom animated/interactive components on screen at any one time, in separate places in the DOM, you might be better off writing your own UI components and making the whole thing into a single Pixi app the way we used to write apps in Flash/Air.

Just to also say: I have one major project that uses Pixi to generate graphics and charts, sometimes up to 100 per page, but it uses one Pixi.Application to do all of that statically, and then copies out temporary images and deploys them around the DOM. When the user clicks on one, a new Pixi app starts there to redraw it as an interactive component. When it loses focus, it's destroyed after awhile. This was my solution to having dozens of Pixi things in an HTML doc at the same time. It's probably not great, but it made sense given the constraints.

**Comment by Mistik** [comment]
Eager for v8 canvas renderer support!

**Comment by brunocapdevila** [comment]
Hi, in our company we are currently stuck with v7 as Chromebook webViews do come with WebGL disabled. We won't be able to upgrade without the fallback to canvas.

**Comment by SmartArray** [comment]
Observing this... Because I am interested in using it when it's out, alternatively I would be willing to add support for the `CanvasRenderer` in the future.

I would love to hear more about the implementation that existed in the past.

I assume filters and shaders were very limited in the previous `CanvasRenderer` implementation.
Also, should there be an option to enable CPU rendering of filters?
I assume one may peek at the existing v7 implementation to pull this task off, right?

**Reply by UlyssesZh** [reply]
> I assume one may peek at the existing v7 implementation to pull this task off, right?

The thing is that v8 greatly revamped the abstractions in the code base, and one needs to understand a lot of both the old code structure and the new code structure in order to port the old implementation. What is worse is that the majority of changes of v8 is in a single enormous commit de4b515.

**Reply by avivost** [reply]
Calling it enormous is an understatement 🤣

**Comment by UlyssesZh** [comment]
@Zyie Could this discussion be converted to an issue? I think this is a very important feature request.

**Comment by bigtimebuddy** [comment, marked as answer]
Quick update to this thread, as of today, we have added experimental Canvas rendering support in [v8.16.0](https://github.com/pixijs/pixijs/releases/tag/v8.16.0)! Huge shout out to @krzys Please let us know what you think and file any bugs or requests as you find them. Thank you all for the feedback and the patience.

**Reply by SmartArray** [reply]
Oh my god, is that real?!

I need to check it out soon. Thanks for your hard work @krzys

**Reply by UlyssesZh** [reply]
I just tried it out. Most things look good!

**Reply by carlos22** [reply]
Same. Some minor glitches but otherwise very cool!

**Reply by hriverahdez** [reply]
Awesome! Thanks for this

**Reply by kirs26** [reply]
Great news! It works with a few minor issues. It would be nice if we didn't have to specify "preference: 'canvas'" so that it would automatically switch to canvas when WebGL isn't available. Thanks!

**Reply by bigtimebuddy** [reply]
Good idea. Please add issues with suggestions or feature requests.

**Reply by kirs26** [reply]
Added: https://github.com/pixijs/pixijs/issues/11869

**Comment by Mistik** [comment]
amazing work @krzys !
